internal/painter/gl/draw.go

```diff
@@ -158,3 +113,67 @@ func (p *glPainter) drawObject(o fyne.CanvasObject, pos fyne.Position, frame fyn
 		p.drawWidget(obj, pos, frame)
 	}
 }
+
+// rectCoords calculates the openGL coordinate space of a rectangle
+func (p *glPainter) rectCoords(size fyne.Size, pos fyne.Position, frame fyne.Size,
+	fill canvas.ImageFill, aspect float32, pad float32) []float32 {
+	size, pos = rectInnerCoords(size, pos, fill, aspect)
+	size, pos = roundToPixelCoords(size, pos, p.pixScale)
+
+	xPos := (pos.X - pad) / frame.Width
+	x1 := -1 + xPos*2
+	x2Pos := (pos.X + size.Width + pad) / frame.Width
+	x2 := -1 + x2Pos*2
+
+	yPos := (pos.Y - pad) / frame.Height
+	y1 := 1 - yPos*2
+	y2Pos := (pos.Y + size.Height + pad) / frame.Height
+	y2 := 1 - y2Pos*2
+
+	return []float32{
+		// coord x, y, z texture x, y
+		x1, y2, 0, 0.0, 1.0, // top left
+		x1, y1, 0, 0.0, 0.0, // bottom left
+		x2, y2, 0, 1.0, 1.0, // top right
+		x2, y1, 0, 1.0, 0.0, // bottom right
+	}
+}
+
+func rectInnerCoords(size fyne.Size, pos fyne.Position, fill canvas.ImageFill, aspect float32) (fyne.Size, fyne.Position) {
+	if fill == canvas.ImageFillContain || fill == canvas.ImageFillOriginal {
+		// change pos and size accordingly
+
```

Comment:
Is this newline intended to be here?

Reply:
This was just moving the code around to fix ordering, it was a separate commit so you can see it was just that ;)